### PR TITLE
refactor(api): add move to well to engine instrument core

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR implements the majority of `ProtocolContext.move_to` on the `ProtocolEngine`-based core. It allows moving to wells and to arbitrary coordinates.

## Changelog

- Update how a `Well` understands its parent `Labware` to use an `__init__` arg rather than flow through the `WellGeometry`
    - This allows `LabwareLike` to be compatible with the new well core without having to implement `WellGeometry` in the engine-core system
    - The `XyzGeometry` classes are mostly redundant with the ProtocolEngine, and I'd like to potentially drop them entirely from the core-classes, if possible
- Implement `InstrumentCore.get_mount`, `get_hardware_state`, and `move_to`
    - `get_mount` and `get_hardware_state` so that... message broker stringification works
    - `move_to` to, you know, move places
- Implement `LabwareCore.get_display_name`
    - It's needed for object `__repr__`s
    - Without `__repr__` working, you straight up can't construct a `LabwareLike` 🤦  

In order to implement `move_to`, I allowed the abstract definition of `move_to` to take a `well_core` argument. The `InstrumentContext` gets the well core, if applicable, out of the `Location.labware`. The legacy core ignores the `well_core` argument, while the engine core uses it to construct the `moveToWell` command parameters. If the legacy core does not receive a `WellCore`, it constructs a `moveToCoordinates`, instead.

## Review requests

Smoke tests with the `enableProtocolEnginePAPICore` FF on and off will be important! Try to make sure movements match between the two systems. To my eyeballs, they move to the same place!

Run smoke tests on #11645 rather than this branch, if possible

### Move to well test protocol

```python
from opentrons import protocol_api, types

metadata = {
    "apiLevel": "2.13",
}


def run(ctx: protocol_api.ProtocolContext) -> None:
    tr = ctx.load_labware("opentrons_96_tiprack_300ul", 1)
    right = ctx.load_instrument("p300_single_gen2", types.Mount.RIGHT, [tr])

    right.move_to(tr.wells()[0].top())
```

### Move to coordinates test protocol

```python
from opentrons import protocol_api, types

metadata = {
    "apiLevel": "2.13",
}


def run(ctx: protocol_api.ProtocolContext) -> None:
    right = ctx.load_instrument("p300_single", types.Mount.RIGHT, [])

    # somewhere in the middle of the deck
    right.move_to(types.Location(point=types.Point(200, 200, 100), labware=None))
```

## Risk assessment

Medium, due to the need to mess around in the current production PAPI code

### CODE REVIEW

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11632** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11632" target="_blank">Graphite</a>  👈
    * **PR #11645** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11645" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11632?utm_source=stack-comment).

File: api/src/opentrons/protocol_runner/task_queue.py [hunk 1]
```diff
@@ -71,10 +71,7 @@ async def _run(self) -> None:
             if self._run_func is not None:
                 await self._run_func()
         except Exception as e:
-            log.debug(
-                "Exception raised during protocol run",
-                exc_info=error,
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/task_queue.py:76]: This was always `None`... oops

File: api/tests/opentrons/data/testosaur_v2.py [hunk 2]
```diff
@@ -10,11 +10,12 @@
...
 def run(ctx: protocol_api.ProtocolContext) -> None:
-    ctx.home()
-    tr = ctx.load_labware("opentrons_96_tiprack_300ul", 1)
-    right = ctx.load_instrument("p300_single", types.Mount.RIGHT, [tr])
-    lw = ctx.load_labware("corning_96_wellplate_360ul_flat", 2)
-    right.pick_up_tip()
-    right.aspirate(10, lw.wells()[0].bottom())
-    right.dispense(10, lw.wells()[1].bottom())
-    right.drop_tip(tr.wells()[-1].top())
+    tr = ctx.load_labware("opentrons_96_tiprack_300ul", 5)
```
Comment (mcous) [inline api/tests/opentrons/data/testosaur_v2.py]: Revert before merge

File: api/src/opentrons/protocol_api/core/engine/instrument.py [hunk 3]
```diff
@@ -107,7 +154,8 @@ def get_available_volume(self) -> float:
         raise NotImplementedError("InstrumentCore.get_available_volume not implemented")
 
     def get_pipette(self) -> PipetteDict:
-        raise NotImplementedError("InstrumentCore.get_pipette not implemented")
+        # TODO(mc, 2022-10-27): implement
+        return cast(PipetteDict, {"display_name": "A pipette"})
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/instrument.py]: Figure out before merge

File: api/src/opentrons/protocol_api/core/protocol_api/well.py [hunk 4]
```diff
@@ -45,10 +44,6 @@ def set_has_tip(self, value: bool) -> None:
         """Set the well as containing or not containing a tip."""
         self._has_tip = value
 
-    def get_display_name(self) -> str:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/well.py]: I might try to revert this particular change if I can

File: api/src/opentrons/protocol_api/core/protocol_api/instrument_context.py [hunk 5]
```diff
@@ -299,3 +299,18 @@ def set_pipette_speed(
             dispense=dispense,
             blow_out=blow_out,
         )
+
+    def flag_unsafe_move(self, location: types.Location) -> None:
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/protocol_api/instrument_context.py:315]: missing docstring?

File: api/src/opentrons/protocol_api/core/protocol_api/instrument_context.py [hunk 6]
```diff
@@ -142,11 +139,14 @@ def home_plunger(self) -> None:
     def move_to(
         self,
         location: types.Location,
+        well_core: Optional[WellImplementation],
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/protocol_api/instrument_context.py:142]: are we silently not using this arg? or will this be implemented in a follow up PR? or is this currently part of `types.Location`?

File: api/src/opentrons/protocol_api/core/protocol_api/legacy_module_core.py [hunk 7]
```diff
@@ -492,6 +497,49 @@ def get_labware_latch_status(self) -> HeaterShakerLabwareLatchStatus:
         """Get the module's labware latch status."""
         return self._sync_module_hardware.labware_latch_status  # type: ignore[no-any-return]
 
+    def flag_unsafe_move(self, to_loc: Location, is_multichannel: bool) -> None:
+        """
+        Raise an error if attempting to perform a move that's deemed unsafe due to
+        the presence of the Heater-Shaker.
+
+        :meta private:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/legacy_module_core.py]: Update docstrings now that it's in the core

File: api/src/opentrons/protocols/api_support/labware_like.py [hunk 8]
```diff
@@ -33,6 +33,8 @@ def __init__(self, labware_like: WrappableLabwareLike):
         self._labware_like = labware_like
         self._type = LabwareLikeType.NONE
 
+        # TODO(mc 2022-10-28): for the love of all that is holy
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocols/api_support/labware_like.py:36]: beautiful comment :-)

File: api/src/opentrons/protocol_api/core/engine/well.py [hunk 9]
```diff
@@ -55,6 +58,30 @@ def get_max_volume(self) -> float:
...
+        """Get the coordinate of the well's bottom, with an z-offset."""
+        return self._engine_client.state.geometry.get_well_position(
+            well_name=self._name,
+            labware_id=self._labware_id,
+            well_location=WellLocation(
+                origin=WellOrigin.BOTTOM, offset=WellOffset(x=0, y=0, z=z_offset)
+            ),
+        )
+
+    def get_center(self) -> Point:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/well.py:84]: Add TODO about why this isn't implemented yet (enum caution)

**Review by TamarZanzouri — APPROVED** [review]
a lot of changes but I think I was able to follow. changes looks good to me! was not tested on hardware but can help with that

**Pull request merged**